README.md
# Bot przypominajka
[![](https://img.shields.io/badge/python-3.7%2B-blue)](https://www.python.org/downloads/release/python-378/) [![](https://img.shields.io/badge/license-MIT-green)](https://opensource.org/licenses/MIT)

Ring to bot umożliwiający przypomnienie nauczycielowi o końcu lekcji. Został wyposażony w funkcję tłumacza google w kilku niezbędnych językach (angielski, rosyjski, hiszpański i niemiecki). Bot potrzebuje uprawnień administratora. Został napisany w języku Python (domyślnie 3.7) na Windowsie, ale na Linuksie powinien działać bez problemów. W razie sugestii lub pytań można udać się do sekcji Issues.

## Komendy bota
**Ogólne**

`//trans <lang> fraza` - przetłumaczenie frazy (język wykrywany automatycznie) na podany

`//ru fraza` - tłumaczenie frazy (język wykrywany automatycznie) na język rosyjski

`//grant <perm> <@użytkownik>` - nadanie uprawnienia dla użytkownika (głosowanie)
- daje uprawnienie kategorii _user_

`//ungrant <perm> <@użytkownik>` - odebranie uprawnienia dla użytkownika (głosowanie)
- próba odebrania uprawnienia kategorii _admin_ skutkuje komunikatem o błędzie

`//perms` - wyświetla permisje autora wiadomości

`//covid` - (beta) wyświetla najnowsze statystyki COVID (źródło: koronawirus.abczdrowie.pl)

`//link` - wysyła link pozwalający botowi dołączyć do serwera

`//pEvent` - (beta) dodaje symbol jakiegoś wydarzenia do avatara (strajk kobiet) (tylko Windows)

**Dla osób z uprawnieniami**

`//ring` - Dołączenie do kanału autora; włączenie sygnału; opuszczenie kanału

`//volume <volume>` - zmiana głośności bota

`//grant <perm> <@użytkownik>` - nadanie uprawnienia dla użytkownika
- _admin_ daje uprawnienia _admin_
- _user_ daje uprawnienia _user_

`//ungrant <perm> <@użytkownik>` - odebranie uprawnienia dla użytkownika
- _user_ może odebrać tylko uprawnienie _user_ (inaczej błąd)

`//votes <votes>` - zmiana ilości osób niezbędnych do przegłosowania uprawnień

`//rm <"msg"> <s> <l>` - (beta, chwilowo dostępne dla każdego użytkownika) wysyła przypomnienie _msg_ po _s_ sekundach na kanał zaczynający się na_l_

## Ostatnia stabilna wersja - v2.4.3
## Ostatnia wersja beta - v2.4.3


additional/changelog.txt
Dzienik zmian (ostatnia stabilna wersja)

2.1.1
Zmiany
- naprowione pozostałości po systemie uprawnień
- obsługa braku argumentu komendy grant

Problemy
- niezidentyfikowany problem słownika

2.1.0
Zmiany
- operowanie na ID

Problemy
- błędy głosowanie (pozostałość nicków)

2.0.1
Zmiany
- komenda votes
- aktualizacja elementów głosowania

Do zrobienia
- operowanie na ID (!!!)

2.0.0
Zmiany
- przetestowane wiele serwerów
- setup

2.0.0-alpha
Zmiany
- działanie na wielu serwerach (nietestowane)
- zmiana długości dzownka na 1,5s

Do zrobienia
- testy
- setup

1.4.3
Zmiany
- zgoda admina na przyznawanie uprawnień
- release

1.4.2
Zmiany
- Zmiana struktury pliku perms.json
- Uporządkowanie przyznawania uprawnień
- Zgoda admina na odbieranie uprawnień
- Zmiana odpowiedzi na komendę perms

Do zrobienia
- Dostosowanie komendy grant
- Utworzenie nowej komendy do zmiany minimum głosów
- komenda do sprawdzenia osób z uprawnieniami admina
- systematyczniej prowadzić README
- utworzyć release

1.4.0
Zmiany
- Dodanie głosowania przy przyznawaniu uprawnienia grant
- użycie elementu discord.Embed

Problemy
- za mało adminofaszyzmu

1.3.1
Zmiany
- Rozdzielenie komendy trans do ru
- Zmiana prefixu bota

1.3.0
Zmiany
- komendy trans
- skrócenie trwania dzwonka do 3 sekund

Problemy
- błędy Google API

1.2.1
Zmiany
- dodanie opisów

1.2.0
Zmiany
- komendy grant i ungrant
- komenda perms

Problemy
- brak opisów
- adminofaszyzm

1.1.0
Zmiany
- próba komendy volume
- usunięcie komendy stop

Problemy
- zarządzanie botem

1.0.0
Zmiany
- komenda ring
- komenda stop

Problemy
- problemy z dołączaniem do kanału
- nie działająca komenda stop
- brak opuszczania kanału


avatars/test.py
import cv2

l_img, s_img = cv2.imread('avatar.png'), cv2.imread('thunder.png', -1)
x_offset, y_offset = 30, 60

y1, y2 = y_offset, y_offset + s_img.shape[0]
x1, x2 = x_offset, x_offset + s_img.shape[1]

alpha_s = s_img[:, :, 3] / 255.0
alpha_l = 1.0 - alpha_s

for c in range(0, 3):
    l_img[y1:y2, x1:x2, c] = (alpha_s * s_img[:, :, c] +
                            alpha_l * l_img[y1:y2, x1:x2, c])

cv2.imwrite("new_avatar.png", l_img)

bot1.py
#https://discord.com/api/oauth2/authorize?client_id=773874155807834113&permissions=8&scope=bot

import asyncio
import discord
import json
from secrets import token1, token2
from discord.ext import commands
from ring import Ring
from trans import Trans
from perms import Perms
from photos import Photos
import datetime
import aiohttp

with open('perms.json') as f:
    data = json.loads(f.read())
dt = datetime.datetime.now

bot = commands.Bot(command_prefix="//", description='End of lesson is here')

@bot.event
async def on_ready():
    print('Logged in as {0} ({0.id})'.format(bot.user))
    print('------')

@bot.command(brief='Displays changelog')
async def changelog(ctx):
    file = discord.File("additional/changelog.txt")
    await ctx.send("Changelog **v2.2.0**", file=file)

@bot.command()
async def covid(ctx):
    if ctx.guild.id == 690261332003651621:
        if ctx.author.id != 723448017151197196:
            await ctx.send('This command is only available for **16 Adrian Lipka**')
            return
    async with ctx.typing():
        async with aiohttp.ClientSession() as cs:
            async with cs.get('https://koronawirus.abczdrowie.pl/#koronawirus-w-polsce') as r:
                resp = await r.text()

        last = resp.find(' dziś')
        new_infections = resp[last-10:last].split('+')[1]

        last -= 13
        infections = resp[last-10:last+10].split('>')[1].split('<')[0]

        embed = discord.Embed(title='COVID statistics', color=0xef3f2d)
        embed.add_field(name="New infections", value=new_infections, inline=True)
        embed.add_field(name="Infections", value=infections, inline=True)
        await ctx.send(embed=embed)

@bot.command()
async def link(ctx):
    await ctx.send('https://discord.com/api/oauth2/authorize?client_id=773874155807834113&permissions=8&scope=bot')

bot.add_cog(Ring(bot, data, dt))
bot.add_cog(Trans(bot, dt))
bot.add_cog(Perms(bot, data, dt))
bot.add_cog(Photos(bot))
bot.run(token1, reconnect=True)


ring.py
import asyncio
import discord
import json
from discord.ext import commands
from random import choice

class Ring(commands.Cog):
    def __init__(self, bot, data, dt):
        self.bot = bot; self.data = data
        self.dt = dt

    def _play(self, ctx, query):
        source = discord.PCMVolumeTransformer(discord.FFmpegPCMAudio(query))
        ctx.voice_client.play(source, after=lambda e: print('Player error: %s' % e) if e else None)

    async def log(self, message):
        await self.bot.get_channel(780912356573184020).send(message)

    @commands.command()
    async def ring(self, ctx):
        serv = str(ctx.guild.id)
        await self.log(f'{self.dt()} RING {ctx.guild.name} {ctx.author}')
        if ctx.author.id in self.data[serv]['perms']['ring']['ba'] or ctx.author.id in self.data[serv]['perms']['ring']['bv']:
            if ctx.author.voice:
                try:
                    ctx.voice_client.disconnect()
                except:
                    pass
                finally:
                    try:
                        await ctx.author.voice.channel.connect()
                    except:
                        await ctx.send('Bot already joined channel. Please wait')
            else:
                await ctx.send("You are not connected to a voice channel.")
                return
            self._play(ctx, 'additional/dzwonek.mp3')
            ctx.voice_client.source.volume = self.data[serv]['vol']/100
            await asyncio.sleep(2)
            await ctx.voice_client.disconnect()
        else:
            await ctx.send('You don\'t have permissions to invoke this command')

    '''@commands.command()
    async def stop(self, ctx):
        if ctx.author.nick[:2] in data['stop']:
            if ctx.voice_client:
                await ctx.voice_client.disconnect()
        else:
            await ctx.send('You don\'t have permissions to invoke this command')'''

    @commands.command(brief='volume <percent> - changes volume')
    async def volume(self, ctx, vol : int):
        serv = str(ctx.guild.id)
        await self.log(f'{self.dt()} VOLUME {ctx.guild.name} {ctx.author}')
        nick = ctx.author.id
        if nick in self.data[serv]['perms']['volume']['ba'] or nick in self.data[serv]['perms']['volume']['bv']:
            self.data[serv]['vol'] = vol
            with open('perms.json', 'w') as f:
                f.write(json.dumps(self.data, indent=2))
            await ctx.send(f'Volume changed to {vol}%')
        else:
            await ctx.send('You don\'t have permissions to invoke this command')

    @commands.command(brief='rm <"message"> <seconds> <starting_letter>')
    async def rm(self, ctx, mess:str, secs: int, start: str, ):
        serv = str(ctx.guild.id)
        await self.log(f'{self.dt()} RM {ctx.guild.name} {ctx.author.name}')
        channels = ctx.guild.channels
        for channel in channels.copy():
            if type(channel) != discord.TextChannel:
                channels.remove(channel)

        def check(msg):
                    return (msg.content.isdigit() or msg.content=='quit') and msg.author.name == ctx.author.name

        if mess == "random":
            if self.data[serv]["phrases"]:
                mess = choice(self.data[serv]["phrases"])
            else:
                await ctx.send("There are **no custom phrases** to choose on this server")
                return
        elif mess == "custom":
            custom = 'Following phrases found:\n```\n'
            if not self.data[serv]['phrases']:
                await ctx.send("There are **no custom phrases** to choose on this server")
                return
            for i, phr in enumerate(self.data[serv]["phrases"]):
                custom += f'{i+1}: "{phr}"\n'
            custom += '```\nType in number of phrase you want to send'
            main = await ctx.send(custom)
            try:
                msg = await self.bot.wait_for('message', timeout=30, check=check)
            except asyncio.TimeoutError:
                await main.edit(':alarm_clock: **Time to choose passed**')
                return
            mess = self.data[serv]['phrases'][int(msg.content)-1]

        if mess not in self.data[serv]["phrases"]:
            self.data[serv]["phrases"].append(mess)
            with open('perms.json', 'w') as file:
                file.write(json.dumps(self.data, indent=2))
        
        if not len(channels):
            await ctx.send(f'There are no text channels')
        else:
            for channel in channels.copy():
                if not channel.name.startswith(start):
                    channels.remove(channel)
            if not len(channels):
                await ctx.send(f'There is no channel that starts with **{start}**')
            elif len(channels) == 1:
                await asyncio.sleep(secs)
                await channels[0].send(mess)
            else:
                possible = 'Possible channels are:\n```\n'
                for i, c in enumerate(channels):
                    possible += f'{i+1}: {c.name}\n'
                possible += '```\nType in number of channel you want to choose'
                main = await ctx.send(possible)

                try:
                    msg = await self.bot.wait_for('message', timeout=30, check=check)
                except asyncio.TimeoutError:
                    await main.edit(':alarm_clock: **Time to choose passed**')
                    return

                await asyncio.sleep(secs)
                await channels[int(msg.content)-1].send(mess)

    @commands.command()
    async def phrase(self, ctx, opt: str, phrase = None):
        serv = str(ctx.guild.id)
        if opt == 'add':
            self.data[serv]['phrases'].append(phrase)
            with open('perms.json', 'w') as file:
                file.write(json.dumps(self.data, indent=2))
            await ctx.send(f'Phrase **{phrase}** added to catalogue')
        elif opt == 'show':
            msg = 'Saved phrases:\n```\n'
            for i, e in enumerate(self.data[serv]['phrases']):
                msg += f'{i+1}: {e}\n'
            msg += '```'
            await ctx.send(msg)
        elif opt == 'remove':
            try:
                self.data[serv]['phrases'].pop(int(phrase)-1)
                with open('perms.json', 'w') as file:
                   file.write(json.dumps(self.data, indent=2))
                await ctx.send(f'Phrase with index {int(phrase)} removed')
            except Exception as e:
                print(e)
                await ctx.send(f'Phrase with index {int(phrase)} does not exist')

trans.py
import asyncio
import discord
from discord.ext import commands
from googletrans import Translator

trns = Translator().translate

class Trans(commands.Cog):
    def __init__(self, bot, dt):
        self.bot = bot; self.dt = dt

    async def log(self, message):
        await self.bot.get_channel(780912356573184020).send(message)

    @commands.command(brief='trans [pl/en/es/ru] "text"')
    async def trans(self, ctx, lang: str, phrase: str):
        serv = ctx.guild.name
        await self.log(f'{self.dt()} TRANS {serv} {ctx.author}')
        try:
            avlb = ['pl', 'en', 'ru', 'es']
            if lang in avlb:
                mess = ''
                trnsd = trns(phrase, dest=lang)
                mess += trnsd.text
                if lang == 'ru':
                    mess += '\nPronunciation: '+trnsd.pronunciation
                await ctx.send(f'Translated text:\n```\n{mess}\n```')
            else:
                await ctx.send('This language is not allowed')
        except Exception as e:
            print(e)
            await ctx.send('This command is temporary unavaliable, because of Google API error')

    @commands.command(brief='ru text')
    async def ru(self, ctx, *args):
        serv = ctx.guild.name
        await self.log(f'{self.dt()} RU {serv} {ctx.author}')
        try:
            trnsd = trns(' '.join(args), dest='ru')
            mess = trnsd.text + '\nPronunciation: '+trnsd.pronunciation
            await ctx.send(f'Translated text:\n```\n{mess}\n```')
        except Exception as e:
            print(e)
            await ctx.send('This command is temporary unavaliable, because of Google API error')


photos.py
from discord.ext import commands
import os
import discord
import cv2

class Photos(commands.Cog):
    def __init__(self, bot):
        self.bot = bot

    @commands.command()
    async def pEvent(self, ctx):
        async with ctx.typing():
            os.system(f'powershell Invoke-WebRequest {ctx.author.avatar_url} -o avatars/avatar.png')

            l_img, s_img = cv2.imread('avatars/avatar.png'), cv2.imread('avatars/thunder.png', -1)
            scale = l_img.shape[1]/256
            dim = (int(s_img.shape[1]*scale), int(s_img.shape[0]*scale))
            print(s_img.shape)
            s_img = cv2.resize(s_img, dim, interpolation=cv2.INTER_AREA)
            print(s_img.shape)
            x_offset, y_offset = 30, 60

            y1, y2 = y_offset, y_offset + s_img.shape[0]
            x1, x2 = x_offset, x_offset + s_img.shape[1]

            alpha_s = s_img[:, :, 3] / 255.0
            alpha_l = 1.0 - alpha_s

            for c in range(0, 3):
                l_img[y1:y2, x1:x2, c] = (alpha_s * s_img[:, :, c] +
                                        alpha_l * l_img[y1:y2, x1:x2, c])

            os.chdir('avatars')
            cv2.imwrite("new_avatar.png", l_img)
            os.chdir('../')

            file = discord.File("avatars/new_avatar.png")
            await ctx.send(file=file)
        


perms.py
import asyncio
import discord
from discord.ext import commands
import json

class Perms(commands.Cog):
    def __init__(self, bot, data, dt):
        self.bot, self.data, self.dt = bot, data, dt

    def saveData(self):
        with open('perms.json', 'w') as file:
            file.write(json.dumps(self.data, indent=2))

    async def log(self, message):
        await self.bot.get_channel(780912356573184020).send(message)

    @commands.Cog.listener()
    async def on_guild_join(self, guild):
        await self.log(f'{self.dt()} Bot joined new guild {guild.name} {guild.id}')
        if guild.me.guild_permissions.administrator:
            channel = await guild.create_text_channel('Ring-setup', overwrites={guild.default_role: discord.PermissionOverwrite(read_messages=False)})
            if str(guild.id) in self.data.keys():
                await channel.send(':white_check_mark: Bot already configured for this server')
                return
            else:
                await channel.send('Hi. Well done. Bot is on the server. **Person with admin rights have to type in `activate` to activate bot**. You have 10 mins')
                def check(msg):
                    if isinstance(msg.author, discord.Member):
                        return (msg.author == guild.owner or msg.author.guild_permissions.administrator) and msg.author != guild.me and msg.channel == channel
                    return False
                while True:
                    try:
                        message = await self.bot.wait_for('message', timeout=600, check=check)
                    except asyncio.TimeoutError:
                        await guild.leave()
                        return

                    if message.content == 'activate':
                        dcp = {"perms":{"ring":{"ba":[message.author.id],"bv":[]}, \
                        "volume":{"ba":[message.author.id], "bv":[]}, \
                        "grant":{"ba":[message.author.id], "bv":[]}}, "vol":100, "votes":10, \
                        "phrases": []}
                        self.data[str(guild.id)] = dcp

                        self.saveData()

                        await channel.send(f':sunglasses: Congratulations. **Bot activated**. User **{message.author.nick}** have admin rights')
                        return

                    else:
                        await channel.send(':x: wrong message')

        else:
            await self.log(f'Bot does not have admin rights {guild.name} ({guild.id})')
            try:
                await guild.text_channels[0].send(':x: **Bot needs admin rights. Add it once again, checking privileges**')
            except discord.Forbidden:
                pass
            await guild.leave()

    @commands.command(brief='grant [perm / help] [@ping ...] - grants permission')
    async def grant(self, ctx, perm : str):
        serv = str(ctx.guild.id)
        await self.log(f'{self.dt()} GRANT {ctx.guild.name} {ctx.author}')

        if perm not in self.data[serv]['perms'].keys():
            await ctx.send('Argument **perm** is missing')
            return

        if ctx.author.id in self.data[serv]['perms']['grant']['ba']:
            if perm == 'help':
                await ctx.send('You can grant following permissions:\n```\nall\ngrant\nring\nvolume```')
                return
            try:
                new = ctx.message.mentions[0].id
            except:
                await ctx.send('You have to mention user')
                return
            if perm=='all':
                for k in self.data[serv]['perms'].keys():
                    if new not in self.data[serv]['perms'][k]['ba']:
                        self.data[serv]['perms'][k]['ba'].append(new)
            else:
                if new not in self.data[serv]['perms'][perm]['ba']:
                    self.data[serv]['perms'][perm]['ba'].append(new)
            self.saveData()
            text = '**__CHANGING PERMISSIONS__**\n' \
                   f'Permission **{perm}** granted for user **{ctx.message.mentions[0].nick}**\n\n' \
                   f':white_check_mark: Done by: {ctx.author}'
            embed = discord.Embed(description=text, color=0xfffffe)
            embed.set_thumbnail(url=ctx.message.mentions[0].avatar_url)
            await ctx.send(embed=embed)
        elif ctx.author.id in self.data[serv]['perms']['grant']['bv']:
            if perm == 'help':
                await ctx.send('You can ungrant following permissions:\n```\ngrant\nring\nvolume\n```')
                return
            new = ctx.message.mentions[0].id
            if not new in self.data[serv]['perms'][perm]['bv']:
                self.data[serv]['perms'][perm]['bv'].append(new)
            self.saveData()
            text = '**__CHANGING PERMISSIONS__**\n' \
                   f'Permission **{perm}** granted for user **{ctx.message.mentions[0].nick}**\n\n' \
                   f':white_check_mark: Done by: {ctx.author}'
            embed = discord.Embed(description=text, color=0xfffffe)
            embed.set_thumbnail(url=ctx.message.mentions[0].avatar_url)
            await ctx.send(embed=embed)
        else:
            new = ctx.message.mentions[0].id; apprs = []
            admin_appr = False; admin_name = 'waiting for approve'; admin_state = ':arrows_counterclockwise:'
            text = '**__CHANGING PERMISSIONS__**\n' \
                   f'Granting permission **{perm}** for user **{ctx.message.mentions[0].nick}**\n\n' \
                   f'{admin_state} Admin: {admin_name}\n' \
                   ':arrows_counterclockwise: Waiting for approve ({}/{})'
            embed = discord.Embed(description=text.format(len(apprs), self.data[serv]['votes']), color=0xfffffe)
            embed.set_thumbnail(url=ctx.message.mentions[0].avatar_url)
            message = await ctx.send(embed=embed)
            await message.add_reaction('✅')

            def check(reaction, user):
                return str(reaction.emoji)=='✅' and reaction.message==message and user!=ctx.guild.me

            while not (not len(apprs)<self.data[serv]['votes'] and admin_appr):
                try:
                    reaction, user = await self.bot.wait_for('reaction_add', timeout=60, check=check)
                except asyncio.TimeoutError:
                    embed = discord.Embed(description=':alarm_clock: **Time to grant passed**')
                    await message.edit(embed=embed)
                    await message.remove_reaction('✅', ctx.guild.me)
                    return

                apprs = await reaction.users().flatten()
                try:
                    apprs.remove(ctx.guild.me)
                except Exception as e:
                    print(e)

                if user.id in self.data[serv]['perms']['grant']['ba']:
                    admin_appr = True
                    admin_name = user
                    admin_state = ':white_check_mark:'

                text = '**__CHANGING PERMISSIONS__**\n' \
                       f'Granting permission **{perm}** for user **{ctx.message.mentions[0].nick}**\n\n' \
                       f'{admin_state} Admin: {admin_name}\n' \
                       ':arrows_counterclockwise: Waiting for approve ({}/{})'
                embed = discord.Embed(description=text.format(len(apprs), self.data[serv]['votes']), color=0xfffffe)
                embed.set_thumbnail(url=ctx.message.mentions[0].avatar_url)
                await message.edit(embed=embed)

            text = '**__CHANGING PERMISSIONS__**\n' \
                   f'Permission **{perm}** granted for user **{ctx.message.mentions[0].id}**\n\n' \
                   f':white_check_mark: Approved by: {admin_name}\n' \
                   ':white_check_mark: Granted in voting (10)'
            embed = discord.Embed(description=text, color=0xfffffe)
            embed.set_thumbnail(url=ctx.message.mentions[0].avatar_url)
            if new not in self.data[serv]['perms'][perm]['bv']:
                self.data[serv]['perms'][perm]['bv'].append(new)
            self.saveData()
            await message.edit(embed=embed)

    @commands.command(brief='ungrant [perm / help] [@ping ...] - takes away permission')
    async def ungrant(self, ctx, perm : str):
        serv = str(ctx.guild.id)
        await self.log(f'{self.dt()} UNGRANT {ctx.guild.name} {ctx.author}')
        if ctx.author.id in self.data[serv]['perms']['grant']['ba']:
            if perm == 'help':
                await ctx.send('You can ungrant following permissions:\n```\nall\ngrant\nring\nvolume\n```')
                return
            new = ctx.message.mentions[0].id
            nick = ctx.message.mentions[0].nick
            if perm=='all':
                for k in self.data[serv]['perms'].keys():
                    if new in self.data[serv]['perms'][k]['ba']:
                        self.data[serv]['perms'][k]['ba'].remove(new)
                    if new in self.data[serv]['perms'][k]['bv']:
                        self.data[serv]['perms'][k]['bv'].remove(new)
            else:
                if new in self.data[serv]['perms'][perm]['ba']:
                    self.data[serv]['perms'][perm]['ba'].remove(new)
                if new in self.data[serv]['perms'][perm]['bv']:
                    self.data[serv]['perms'][perm]['bv'].remove(new)
            self.saveData()
            text = '**__CHANGING PERMISSIONS__**\n' \
                   f'Permission **{perm}** taken away from user **{nick}**\n\n' \
                   f':white_check_mark: Done by: {ctx.author}'
            embed = discord.Embed(description=text, color=0xfffffe)
            embed.set_thumbnail(url=ctx.message.mentions[0].avatar_url)
            await ctx.send(embed=embed)
        elif ctx.author.id in self.data[serv]['perms']['grant']['bv']:
            if perm == 'help':
                await ctx.send('You can ungrant following permissions:\n```\ngrant\nring\nvolume\n```')
                return
            new = ctx.message.mentions[0].id
            nick = ctx.message.mentions[0].nick
            if new in self.data[serv]['perms'][perm]['ba']:
                text = '**__CHANGING PERMISSIONS__**\n' \
                       f'You cannot take away this permission from user **{nick}**\n\n' \
                       'Only admin feature :shrug: :sunglasses:'
                embed = discord.Embed(description=text, color=0xfffffe)
                embed.set_thumbnail(url=ctx.message.mentions[0].avatar_url)
                message = await ctx.send(embed=embed)
                return
            if new in self.data[serv]['perms'][perm]['bv']:
                self.data[serv]['perms'][perm]['bv'].remove(new)
            self.saveData()
            text = '**__CHANGING PERMISSIONS__**\n' \
                   f'Permission **{perm}** taken away from user **{nick}**\n\n' \
                   f':white_check_mark: Done by: {ctx.author}'
            embed = discord.Embed(description=text, color=0xfffffe)
            embed.set_thumbnail(url=ctx.message.mentions[0].avatar_url)
            await ctx.send(embed=embed)
        else:
            new = ctx.message.mentions[0].id
            nick = ctx.message.mentions[0].nick
            if new in self.data[serv]['perms'][perm]['ba']:
                text = '**__CHANGING PERMISSIONS__**\n' \
                       f'You cannot take away this permission from user **{nick}**\n\n' \
                       'Only admin feature :shrug: :sunglasses:'
                embed = discord.Embed(description=text, color=0xfffffe)
                embed.set_thumbnail(url=ctx.message.mentions[0].avatar_url)
                message = await ctx.send(embed=embed)
                return
            apprs = []; admin_appr = False; admin_name = 'waiting for approve'; admin_state = ':arrows_counterclockwise:'
            text = '**__CHANGING PERMISSIONS__**\n' \
                   f'Taking away permission **{perm}** from user **{nick}**\n\n' \
                   f'{admin_state} Admin: {admin_name}\n' \
                   ':arrows_counterclockwise: Waiting for approve ({}/10)'
            embed = discord.Embed(description=text.format(len(apprs)), color=0xfffffe)
            embed.set_thumbnail(url=ctx.message.mentions[0].avatar_url)
            message = await ctx.send(embed=embed)
            await message.add_reaction('✅')

            def check(reaction, user):
                return str(reaction.emoji)=='✅' and reaction.message==message and user!=ctx.guild.me

            while not (not len(apprs)<self.data[serv]['votes'] and admin_appr):
                try:
                    reaction, user = await self.bot.wait_for('reaction_add', timeout=60, check=check)
                except asyncio.TimeoutError:
                    embed = discord.Embed(description=':alarm_clock: **Time to ungrant passed**')
                    await message.edit(embed=embed)
                    await message.remove_reaction('✅', ctx.guild.me)
                    return

                apprs = await reaction.users().flatten()
                try:
                    apprs.remove(ctx.guild.me)
                except Exception as e:
                    print(e)

                if user.id in self.data[serv]['perms']['grant']['ba']:
                    admin_appr = True
                    admin_name = user
                    admin_state = ':white_check_mark:'

                text = '**__CHANGING PERMISSIONS__**\n' \
                       f'Taking away permission **{perm}** from user **{nick}**\n\n' \
                       f'{admin_state} Admin: {admin_name}\n' \
                       ':arrows_counterclockwise: Waiting for approve ({}/10)'
                embed = discord.Embed(description=text.format(len(apprs)), color=0xfffffe)
                embed.set_thumbnail(url=ctx.message.mentions[0].avatar_url)
                await message.edit(embed=embed)

            text = '**__CHANGING PERMISSIONS__**\n' \
                   f'Permission **{perm}** taken away from user **{nick}**\n\n' \
                   f':white_check_mark: Approved by: {admin_name}\n' \
                   ':white_check_mark: Ungranted in voting (10)'
            embed = discord.Embed(description=text, color=0xfffffe)
            embed.set_thumbnail(url=ctx.message.mentions[0].avatar_url)
            if new in self.data[serv]['perms'][perm]['bv']:
                self.data[serv]['perms'][perm]['bv'].remove(new)
            self.saveData()
            await message.edit(embed=embed)

    @commands.command(brief='Check permissions')
    async def perms(self, ctx):
        serv = str(ctx.guild.id)
        await self.log(f'{self.dt()} PERMS {ctx.guild.name} {ctx.author}')
        nick = ctx.author.id
        mess = 'Your permissions:\n```\nBy admin:\n'
        admin = ''; voting = ''
        for k in self.data[serv]['perms'].keys():
            if nick in self.data[serv]['perms'][k]['ba']:
                admin = admin + f'{k}\n'
            if nick in self.data[serv]['perms'][k]['bv']:
                voting = voting + f'{k}\n'
        if admin:
            mess += admin
            if ctx.author.id == 723448017151197196:
                mess += 'masterOfCovid\n'
        else:
            mess += 'None\n'
        mess += '\nBy voting or by privileged user:\n'
        if voting:
            mess += voting
        else:
            mess += 'None\n'
        mess += '\n```'
        await ctx.send(mess)

    @commands.command(brief='votes <votes> - changes votes limit')
    async def votes(self, ctx, votes: int):
        serv = str(ctx.guild.id)
        await self.log(f'{self.dt()} VOTES {ctx.guild.name} {ctx.author}')
        if ctx.author.id in self.data[serv]['perms']['grant']['ba']:
            self.data[serv]['votes'] = votes
            self.saveData()
            await ctx.send(f'Votes limit changed to **{votes}**')
        elif ctx.author.id in self.data[serv]['perms']['grant']['bv']:
            await ctx.send('You can not invoke this command. Only admin feature :shrug: :sunglasses:')
        else:
            await ctx.send('You do not have permission to invoke this command')
